infrastructure/tari_util/src/chacha20.rs

```diff
@@ -20,122 +20,220 @@
+        let mut cipher_bytes: Vec<u8> = Vec::with_capacity(block_byte_count as usize);
+        for counter in 1..=block_count as u32 {
+            let mut state = Self::construct_state(key, nonce, counter);
+            let cipher_block = Self::chacha20_block(&state);
+            state.clear();
+            // convert cipher block to bytes
+            let mut block_bytes: Vec<u8> = Vec::with_capacity(BYTES_PER_BLOCK);
+            for block in cipher_block.iter() {
+                block_bytes.append(&mut block.to_ne_bytes().to_vec())
+            }
+            cipher_bytes.append(&mut block_bytes)
+        }
+        (cipher_bytes)
+    }
 
-/// Encode the provided input bytes using a chacha20 keystream
-pub fn encode_with_nonce(bytes: &[u8], key: &[u8; 32], nonce: &[u32; 3]) -> Vec<u8> {
-    const BYTES_PER_BLOCK: usize = 64;
-    let block_count = (bytes.len() as f64 / BYTES_PER_BLOCK as f64).ceil() as usize;
-    let cipher_bytes = chacha20_cipher_keystream(key, nonce, block_count);
-    let mut encoded_bytes = Vec::with_capacity(bytes.len());
-    for i in 0..bytes.len() {
-        encoded_bytes.push(cipher_bytes[i] ^ bytes[i]);
+    /// Encode the provided input bytes using a chacha20 keystream
+    fn encode_with_nonce(bytes: &[u8], key: &[u8; 32], nonce: &[u8; 12]) -> Vec<u8> {
+        const BYTES_PER_BLOCK: usize = 64;
+        let block_count = (bytes.len() as f64 / BYTES_PER_BLOCK as f64).ceil() as usize;
+        let cipher_bytes = Self::chacha20_cipher_keystream(key, nonce, block_count);
+        let mut encoded_bytes = Vec::with_capacity(bytes.len());
+        for i in 0..bytes.len() {
+            encoded_bytes.push(cipher_bytes[i] ^ bytes[i]);
+        }
+        (encoded_bytes)
     }
-    (encoded_bytes)
-}
 
-/// Decode the provided input bytes using a chacha20 keystream and a default nonce
-pub fn decode(bytes: &[u8], key: &[u8; 32]) -> Vec<u8> {
-    (encode(bytes, &key))
+    /// Decode the provided input bytes using a chacha20 keystream
+    fn decode_with_nonce(bytes: &[u8], key: &[u8; 32], nonce: &[u8; 12]) -> Vec<u8> {
+        (Self::encode_with_nonce(bytes, &key, &nonce))
+    }
 }
 
-/// Decode the provided input bytes using a chacha20 keystream
-pub fn decode_with_nonce(bytes: &[u8], key: &[u8; 32], nonce: &[u32; 3]) -> Vec<u8> {
-    (encode_with_nonce(bytes, &key, &nonce))
+impl<D> Cipher<D> for ChaCha20
+where D: ByteArray
+{
+    fn seal(plain_text: &D, key: &Vec<u8>, nonce: &Vec<u8>) -> Result<Vec<u8>, CipherError> {
+        // Validation
+        if key.len() != 32 {
+            return Err(CipherError::KeyLengthError);
+        }
+        if nonce.len() != 12 {
+            return Err(CipherError::NonceLengthError);
+        }
+
+        let mut sized_key = [0; 32];
+        // sized_key.copy_from_slice(&key[..32]);
+        sized_key.copy_from_slice(key.as_slice());
+
+        let mut sized_nonce = [0; 12];
+        // sized_nonce.copy_from_slice(&nonce[..12]);
```

Comment:
You can probably remove the comments from ln 151 and 1550

Reply:
Oh woops, yes I can